server/src/main/java/org/apache/druid/segment/loading/LeastBytesUsedStorageLocationSelectorStrategy.java

```diff
@@ -36,7 +38,13 @@
 
   private List<StorageLocation> storageLocations;
 
-  public LeastBytesUsedStorageLocationSelectorStrategy(List<StorageLocation> storageLocations)
+  @JsonCreator
+  public LeastBytesUsedStorageLocationSelectorStrategy()
+  {
+  }
+
+  @VisibleForTesting
+  LeastBytesUsedStorageLocationSelectorStrategy(List<StorageLocation> storageLocations)
```

Comment:
I think this approach is brittle and might break when a new implementation of `StorageLocationSelectorStrategy` is added. Instead, I think we should make `List<StorageLocation> storageLocations` injectable instead via a module, something like

```
@Provides
@Singleton
public List<StorageLocation> provideStorageLocations(SegmentLoaderConfig config)
{
  this.locations = new ArrayList<>();
  for (StorageLocationConfig locationConfig : config.getLocations()) {
    locations.add(
        new StorageLocation(
            locationConfig.getPath(),
            locationConfig.getMaxSize(),
            locationConfig.getFreeSpacePercent()
        )
    );
return locations;
}
```

Reply:
Good idea.

Reply:
Hi @suneet-s I check the code and find that this approach is a little bit complex for the existing code.

Storage selector strategy object is constructed during deserialization of `SegmentLoaderConfig`,  and only after the construction of `SegmentLoaderConfig`, could it be possible to inject `SegmentLoaderConfig` to other objects to get list of `StorageLocationConfig` inside the object.

If we want to inject `StorageLocation` objects as the way you suggest, the strategy object must be separated from `SegmentLoaderConfig` into another new config class so that both `SegmentLoaderConfig` and this new config class can be injected into `SegmentLocalCacheManager`.  There will be lots of test cases involved to change to meet the new ctor of `SegmentLocalCacheManager`. So I think these change involve more complexity.

Back to the concern you mentioned, I think there's no need to worry that new implementation would break the constraints. Because if it breaks,  the problem would be easily detected during unit test or integrated test. 

What do you think ?